js/eventloop.go

```diff
@@ -0,0 +1,127 @@
+/*
+ *
+ * k6 - a next-generation load testing tool
+ * Copyright (C) 2021 Load Impact
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Affero General Public License as
+ * published by the Free Software Foundation, either version 3 of the
+ * License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package js
+
+import (
+	"context"
+	"sync"
+)
+
+// an event loop
+// TODO: DO NOT USE AS IT'S NOT DONE
+type eventLoop struct {
+	queueLock     sync.Mutex
+	queue         []func()
+	started       int
+	wakeupCh      chan struct{} // maybe use sync.Cond ?
+	reservedCount int
+}
+
+func newEventLoop() *eventLoop {
+	return &eventLoop{
+		wakeupCh: make(chan struct{}, 1),
+	}
+}
+
+// RunOnLoop queues the function to be called from/on the loop
+// This needs to be called before calling `Start`
+// TODO maybe have only Reserve as this is equal to `e.Reserve()(f)`
+func (e *eventLoop) RunOnLoop(f func()) {
+	e.queueLock.Lock()
+	e.queue = append(e.queue, f)
+	e.queueLock.Unlock()
+	select {
+	case e.wakeupCh <- struct{}{}:
+	default:
+	}
+}
+
+// Reserve "reserves" a spot on the loop, preventing it from returning/finishing. The returning function will queue it's
+// argument and wakeup the loop if needed and also unreserve the spot so that the loop can exit.
+// this should be used instead of MakeHandledPromise if a promise will not be returned
+// TODO better name
+func (e *eventLoop) Reserve() func(func()) {
+	e.queueLock.Lock()
+	e.reservedCount++
+	started := e.started
+	e.queueLock.Unlock()
+
+	return func(f func()) {
+		e.queueLock.Lock()
+		if started != e.started {
+			e.queueLock.Unlock()
+			return
```

Comment:
> This will silently discard f if Reserve was called before Start, right?

Yes it will, which likely *can* be fixed by increasing `started` at the end instead of at the beginning :shrug: 

But to be honest I don't think this is a problem. 
eventLoop will always be only internally used in the runner/bundle/vu and as such it has very little actual places where we directly interact with it. 

> Could this info be added somewhere in this file, as a comment?

https://github.com/grafana/k6/pull/2228/files#diff-da0f214f9c82c517bd4fe1f550a32b3c37b378419325e2cce650f5fcbedf73f0R45
But the TODO below is now ... less viable.

> Given this complex rules to how to use eventLoop, I don't think it's a good idea to re-export / expose any of these methods of eventLoop to the module level interface as they are sweat_smile

This has never been the idea if anything as I mentioned I want to only export `Reserve`. The actual module developers can only want to add things to the queue and re-exporting Reserve for them seems to give both the most versatility and we can document that you need to call it before you make a goroutine and return from a function that will return to goja.

```
func (m *moduleInstance) SomethingThatIsCalledOnTheLoopForExampleJustACallMadeFromJS(...) someObject {
  addToLoop := m.vu.ReserveOnLoop() // will need better name
  go func() {
     // do something
     addToLoop(someThingToBeCalledOnTheLoop) // this is a function and it can use the runtime to call some more js code
  }()
  return someObject{} // this can be a Promise where the above `someThingToBeCalledOnTheLoop` either resolves or rejects it.
}
```
Here the Loop is definitely started and if it gets stopped in between ReserveOnLoop and `addToLoop` being called we definitely don't want this to be added to the next queue (even if the eventLoop has been restarted, for example for a new iteration ;) )

A possible redoing of the current API is to have:
1. remove RunOnLoop entirely
2. have Start to take a function that should be queued before executing
3. no changes to Reserve
Which I am now going to try out

Reply:
While doing some related changes I somewhat feel that 
1. we can recreate the eventLoop on each iteration which might help with the other problems (and complexity)
2. It might actually be better if we don't stop on a context or at least not on the same one that is used for everything else. The "main" context also triggers an interrupt on the goja machine so it means the loop should clear either way and as long as that can't be overrun in js I do think modules should probably have a way to queue something on the queue when the scenario is finished, if that is what they want/need. 

This, unfortunately, goes more in the direction of iteration/scenario/vu lifecycle hooks for extensions part and I think that problem will likely need a lot more work, so I will try out some stuff and see whether I can remove more stuff from the current code and simplify it even more

Reply:
> 1. remove RunOnLoop entirely
> 2. have Start to take a function that should be queued before executing
> 3. no changes to Reserve

:+1: yes, that might be a good alternative